core/storage/translation/gae_models.py

```diff
@@ -0,0 +1,147 @@
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for machine translation."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.platform import models
+import utils
+
+(base_models,) = models.Registry.import_models(
+    [models.NAMES.base_model])
+
+datastore_services = models.Registry.import_datastore_services()
+
+
+class MachineTranslatedTextModel(base_models.BaseModel):
+    """Model for storing machine translations. Model instances have a key
+    generated from the source and target language codes, followed by a sha-1
+    hash of the origin text formated as follows:
+
+        [source_language_code]:[target_language_code]:[hashed_origin_text]
+
+    See MachineTranslatedTextModel._generate_id() below for details."""
+
+    # The untranslated text.
+    origin_text = datastore_services.TextProperty(required=True, indexed=False)
+    # An sha-1 hash of the origin text.
+    hashed_origin_text = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the untranslated text.
+    source_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the translation.
+    target_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The translation.
+    translated_text = datastore_services.TextProperty(
+        required=True, indexed=False)
+
+    @classmethod
+    def create(
+            cls, source_language_code, target_language_code, origin_text,
+            translated_text):
+        """Creates a new MachineTranslatedTextModel instance and returns its
+        ID.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            origin_text: str. The untranslated text.
+            translated_text: str. The translation.
+
+        Returns:
+            str. The id of the newly created MachineTranslatedTextInstance.
+        """
+        hashed_origin_text = utils.convert_to_hash(origin_text, 50)
+        entity_id = cls._generate_id(
+            source_language_code, target_language_code, hashed_origin_text)
+        translation_entity = cls(
+            id=entity_id,
+            hashed_origin_text=hashed_origin_text,
+            source_language_code=source_language_code,
+            target_language_code=target_language_code,
+            origin_text=origin_text,
+            translated_text=translated_text)
+        translation_entity.put()
+        return entity_id
+
+    @classmethod
```

Comment:
Not exactly sure why we need this decorator here. I see this being used in other `_generate_id` methods in the codebase, but it doesn't really seem too add much value especially since it's a "private" method anyway.

Reply:
I changed this to be a static method since it's agnostic to model instances.